src/Simplex/Messaging/Agent/Client.hs

```diff
@@ -714,13 +729,60 @@ runSMPServerTest c userId (ProtoServerWithAuth srv auth) = do
         ok <- tcpTimeout (networkConfig cfg) `timeout` closeProtocolClient smp
-        incClientStat c userId smp "TEST" "OK"
-        pure $ either Just (const Nothing) r <|> maybe (Just (SMPTestFailure TSDisconnect $ BROKER addr TIMEOUT)) (const Nothing) ok
+        incClientStat c userId smp "SMP_TEST" "OK"
+        pure $ either Just (const Nothing) r <|> maybe (Just (ProtocolTestFailure TSDisconnect $ BROKER addr TIMEOUT)) (const Nothing) ok
       Left e -> pure (Just $ testErr TSConnect e)
   where
     addr = B.unpack $ strEncode srv
-    testErr :: SMPTestStep -> SMPClientError -> SMPTestFailure
-    testErr step = SMPTestFailure step . protocolClientError SMP addr
+    testErr :: ProtocolTestStep -> SMPClientError -> ProtocolTestFailure
+    testErr step = ProtocolTestFailure step . protocolClientError SMP addr
+
+runXFTPServerTest :: forall m. AgentMonad m => AgentClient -> UserId -> XFTPServerWithAuth -> m (Maybe ProtocolTestFailure)
+runXFTPServerTest c userId (ProtoServerWithAuth srv auth) = do
+  cfg <- asks $ xftpCfg . config
+  xftpNetworkConfig <- readTVarIO $ useNetworkConfig c
+  workDir <- getXFTPWorkPath
+  filePath <- getTempFilePath workDir
+  rcvPath <- getTempFilePath workDir
+  liftIO $ do
+    let tSess = (userId, srv, Nothing)
+    X.getXFTPClient tSess cfg {xftpNetworkConfig} (\_ -> pure ()) >>= \case
+      Right xftp -> do
+        (sndKey, spKey) <- liftIO $ C.generateSignatureKeyPair C.SEd25519
+        (rcvKey, rpKey) <- liftIO $ C.generateSignatureKeyPair C.SEd25519
+        createTestChunk filePath
+        digest <- liftIO $ C.sha256Hash <$> B.readFile filePath
+        let file = FileInfo {sndKey, size = chSize, digest}
+            chunkSpec = X.XFTPChunkSpec {filePath, chunkOffset = 0, chunkSize = chSize}
+        r <- runExceptT $ do
+          (sId, [rId]) <- liftError (testErr TSCreateFile) $ X.createXFTPChunk xftp spKey file [rcvKey] auth
+          liftError (testErr TSUploadFile) $ X.uploadXFTPChunk xftp spKey sId chunkSpec
+          liftError (testErr TSDownloadFile) $ X.downloadXFTPChunk xftp rpKey rId $ XFTPRcvChunkSpec rcvPath chSize digest
+          rcvDigest <- liftIO $ C.sha256Hash <$> B.readFile rcvPath
+          unless (digest == rcvDigest) $ throwError $ ProtocolTestFailure TSCompareFile $ XFTP DIGEST
+          liftError (testErr TSDeleteFile) $ X.deleteXFTPChunk xftp spKey sId
+        ok <- tcpTimeout xftpNetworkConfig `timeout` X.closeXFTPClient xftp
+        incClientStat c userId xftp "XFTP_TEST" "OK"
+        pure $ either Just (const Nothing) r <|> maybe (Just (ProtocolTestFailure TSDisconnect $ BROKER addr TIMEOUT)) (const Nothing) ok
+      Left e -> pure (Just $ testErr TSConnect e)
+  where
+    addr = B.unpack $ strEncode srv
+    testErr :: ProtocolTestStep -> XFTPClientError -> ProtocolTestFailure
+    testErr step = ProtocolTestFailure step . protocolClientError XFTP addr
+    chSize :: Integral a => a
+    chSize = kb 256
+    getTempFilePath :: FilePath -> m FilePath
```

Comment:
maybe make a utility function of it to re-use in xftp receive, and later send?

Reply:
or maybe better to do it later, when send is done

Reply:
I looked at it, and it uses different extension, yes - can be refactored out later.